python/ray/train/_internal/checkpoint.py

```diff
@@ -102,14 +102,27 @@ def _process_checkpoint(
         """Ray Train entrypoint. Perform all processing for a checkpoint."""
         # Get checkpoint from first worker.
         checkpoint_data = checkpoint_results[0].data
-
-        # Decode checkpoint.
-        checkpoint_data = decode_checkpoint_fn(checkpoint_data)
+        checkpoint_metadata = checkpoint_results[0].metadata or {}
+        encoded = checkpoint_results[0].encoded
+
+        # Decode checkpoint if it's not persisted
+        if encoded:
+            checkpoint_data = decode_checkpoint_fn(checkpoint_data.to_dict())
+            checkpoint_data_dict = checkpoint_data
+        # This is too risky for now (will be saved to a tmp dir)
+        # elif checkpoint_data.uri:
+        #     # TODO: ensure that the dir is created in the proper place
+        #     checkpoint_data = checkpoint_data.to_directory()
+        #     checkpoint_data_dict = {}
```

Comment:
Ah I don't think this case is necessary at all. If decode_checkpoint_fn is only there for legacy reasons, then the Checkpoint will always be a dictionary checkpoint.